compiler/hash-typecheck/src/ops/discover.rs

```diff
@@ -0,0 +1,1078 @@
+    /// given [HashSet].
+    pub(crate) fn add_free_sub_vars_in_level0_term_to_set(
+        &self,
+        term: &Level0Term,
+        result: &mut HashSet<SubVar>,
+    ) {
+        match term {
+            Level0Term::Rt(ty_term_id) => {
+                self.add_free_sub_vars_in_term_to_set(*ty_term_id, result)
+            }
+            Level0Term::FnLit(fn_lit) => {
+                // Forward to fn type and return value
+                self.add_free_sub_vars_in_term_to_set(fn_lit.fn_ty, result);
+                self.add_free_sub_vars_in_term_to_set(fn_lit.return_value, result);
+            }
+            Level0Term::FnCall(fn_call) => {
+                // Forward to subject and args:
+                self.add_free_sub_vars_in_term_to_set(fn_call.subject, result);
+                self.add_free_sub_vars_in_args_to_set(fn_call.args, result);
+            }
+            Level0Term::Tuple(tuple_lit) => {
+                self.add_free_sub_vars_in_args_to_set(tuple_lit.members, result);
+            }
+            Level0Term::Constructed(constructed) => {
+                self.add_free_sub_vars_in_term_to_set(constructed.subject, result);
+                self.add_free_sub_vars_in_args_to_set(constructed.members, result);
+            }
+            Level0Term::EnumVariant(_) | Level0Term::Lit(_) => {}
+        }
+    }
+
+    /// Add the free variables that exist in the given [Level1Term], to the
+    /// given [HashSet].
+    pub(crate) fn add_free_sub_vars_in_level1_term_to_set(
+        &self,
+        term: &Level1Term,
+        result: &mut HashSet<SubVar>,
+    ) {
+        match term {
+            Level1Term::ModDef(_) | Level1Term::NominalDef(_) => {}
+            Level1Term::Tuple(tuple_ty) => {
+                // Add the free variables in the parameters (don't remove the parameter names)
+                self.add_free_sub_vars_in_params_to_set(tuple_ty.members, result);
+            }
+            Level1Term::Fn(fn_ty) => {
+                // Add the free variables in the parameters and return type.
+                self.add_free_sub_vars_in_params_to_set(fn_ty.params, result);
+                self.add_free_sub_vars_in_term_to_set(fn_ty.return_ty, result);
+            }
+        }
+    }
+
+    /// Add the free variables that exist in the given [Level2Term], to the
+    /// given [HashSet].
+    pub(crate) fn add_free_sub_vars_in_level2_term_to_set(
+        &self,
+        term: &Level2Term,
+        _result: &mut HashSet<SubVar>,
+    ) {
+        match term {
+            Level2Term::Trt(_) | Level2Term::AnyTy => {}
+        }
+    }
+
+    /// Add the free variables that exist in the given [Level3Term], to the
+    /// given [HashSet].
+    pub(crate) fn add_free_sub_vars_in_level3_term_to_set(
+        &self,
+        term: &Level3Term,
+        _: &mut HashSet<SubVar>,
+    ) {
+        match term {
+            Level3Term::TrtKind => {}
+        }
+    }
```

Comment:
these are still to-dos?

Reply:
no

Reply:
As in, the actual impl for these is as it is?

Reply:
Yes, fixed.